### ISSUE TITLE
Enforce JS style guide (again)

### ISSUE THREAD

**Issue by gasman** [wagtail/wagtail#2797]
As reported in https://groups.google.com/d/msg/wagtail-developers/MjNuJnhXJv4/cPZX5LhnBgAJ - it appears that JS linting was broken when #2275 was merged (`npm run lint:js` now only checks the `client` folder). We need to:
- fix the task to run across the whole codebase
- add it to the `make lint` task https://github.com/torchbox/wagtail/blob/master/Makefile#L18-L20
- fix the inevitable style violations that have crept in since #1237
- add it to our CI setup

**Comment by thibaudcolas** [comment]
I tried this very briefly yesterday, works fine but it raises upwards of 5000 linting issues.

Waiting for the next release of [Prettier](https://prettier.io) (due sometime over the next few weeks, https://github.com/prettier/prettier/issues/2589), which has interesting config options, and could allow us to reformat all of the files in a sustainable way.

**Comment by lb-** [comment]
@thibaudcolas do you know of any progress on this one since you looked at this in 2017?

Prettier and Eslint have both had substantial updates in the last few years, using them together and help a lot. Eslint's fix option in conjunction with prettier can make a lot of minor linting issues automated.

I'm up to help with this if needed, I have implemented pretter/eslint configs on a few projects.

If we want to look in to pre-commit hooks, Husky seems to be a good option.

**Comment by thibaudcolas** [comment]
Nope, there's been no progress that I'm aware of. One of the problems to solve here is that files in `client` are transpiled with Babel, and bundled with Webpack, while the rest of the JS isn't. This means we'd need separate ESLint/Prettier configs, so we don't treat non-transpiled code as if it was  ES2015+.

I don't think this is particularly blocking though – main reason why this hasn't happened is that it's not particularly appealing to fix Wagtail's old JS.

**Comment by lb-** [comment]
@thibaudcolas - makes sense.

Hopefully this will become a bit easier when all code is moved to the webpack build process (client folder)

cc @jonnyscholes

**Comment by thibaudcolas** [comment]
See #5252. I think once that's done this JS linting enforcement issue will be 99% resolved.

**Comment by jonnyscholes** [comment]
Yep I'm linting the new JS as it goes in so in theory once that's all done all our JS will be consistent 😁

Although we might have to make some concessions for some jQuery plugins that whilst I'm hoping to remove, might have to stay around for a bit longer. Trying to avoid that though!

**Comment by lb-** [comment]
@jonnyscholes  - epic stuff, I am hoping to get some time to look at our external eslint config and get that upgraded :) 

let me know if you need a hand with some of this stuff.

**Comment by thibaudcolas** [comment]
Following up a year later! 😄 

https://github.com/wagtail/eslint-config-wagtail is now up-to-date with the Airbnb config and ESLint v6, thanks to work by @lb-. This is now released as `@wagtail/eslint-config-wagtail`. Next step here is to update the Wagtail codebase to use that config – I tried briefly and this raised lots of errors in `client`, which mostly seemed justified if we want to stick with quite a strict config.

---

For the JS that’s under different apps in `wagtail/` – ESLint supports nested configurations. I would welcome a PR that adds a separate ESLint configuration file `.eslintrc.js` inside `wagtail`, overriding the configuration specifically for apps that are in there. I find it inelegant to have a separate config file in there, but pragmatically it would be useful for us to have _some_ form of linting on those files, and they do need their own configuration. This config file will need to:

- Lint for ES5 JS, while code under `client/` can use ES2015+ syntax and APIs
- Have configuration for a number of global variables used throughout this code (e.g. jQuery)

Otherwise I would welcome it being as lean as possible (extending `eslint:recommended` and then disabling as few rules as possible) – if there are rules that are only infringed on in a few places, I’d rather have local overrides with `eslint-disable` comments.
